### PR TITLE
Add chat template from HF tokenizer

### PR DESCRIPTION
**Pull request by cuichenx**
# What does this PR do ?
This PR adds support for applying HuggingFace tokenizer chat templates during dataset preprocessing for supervised fine-tuning (SFT). When enabled, the tokenizer's apply_chat_template method is used to format input/output pairs according to the model's expected chat format, ensuring proper special tokens and formatting are applied during data preprocessing.

# Changelog

- Add specific line by line info of high level changes in this PR.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci)in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information
Two runs comparing finetuning with and without chat template. It's expected to see different initial loss but similar converged loss.
image

- Related to # (issue)

### CODE REVIEW

File: src/megatron/bridge/data/hf_processors/squad.py [hunk 1]
```diff
@@ -50,6 +56,8 @@ def process_squad_example(
     """
     # Format input as: "Context: ... Question: ... Answer:"
     _input = f"Context: {example['context']} Question: {example['question']} Answer:"
+    if apply_tokenizer_chat_template and tokenizer is not None and tokenizer.library == "huggingface":
+        _input = tokenizer.apply_chat_template([{"role": "user", "content": _input}], tokenize=False)
```
Comment (liding-nv) [inline src/megatron/bridge/data/hf_processors/squad.py:60]: Need to check if tokenizer has `chat_template` or not.
Base model typically does not have chat template, and it will raise an error when you run `tokenizer.apply_chat_template`

File: src/megatron/bridge/data/hf_processors/squad.py [hunk 2]
```diff
@@ -50,6 +56,8 @@ def process_squad_example(
...
 
     # Use the first answer as the primary output
     _output = example["answers"]["text"][0]
```
Comment (liding-nv) [inline src/megatron/bridge/data/hf_processors/squad.py:63]: Q: do we need to process output as well?
If the sample is
```
{"Context": XXXX, "Question": YYYY, "Answer": ZZZZ}
```
Then we construct the chat as 
```
[
{"role": "user", "content": "Context: XXXX Question: YYYY Answer"}, 
{"role": "assistant", "content": "ZZZZ"}
]
```
After chat template, it becomes
```
<|im_start|>system
<|im_end|>
<|im_start|>user
Context: XXXX Question: YYYY Answer<|im_end|>
<|im_start|>assistant
ZZZZ<|im_end|>
```
So output part also has special tokens around it

File: src/megatron/bridge/training/tokenizers/utils.py [hunk 3]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from megatron.core.tokenizers import MegatronTokenizer
+from megatron.core.tokenizers import MegatronTokenizer, MegatronTokenizerBase
 
 from megatron.bridge.training.tokenizers.config import TokenizerConfig
```
Comment (liding-nv) [inline src/megatron/bridge/training/tokenizers/utils.py:18]: use `TYPE_CHECKING` here?

File: src/megatron/bridge/training/tokenizers/tokenizer.py [hunk 4]
```diff
@@ -7,7 +7,7 @@
 from pathlib import Path
 from typing import Dict, List, Optional
 
-from megatron.core.tokenizers import MegatronTokenizer
+from megatron.core.tokenizers import MegatronTokenizer, MegatronTokenizerBase
```
Comment (liding-nv) [inline src/megatron/bridge/training/tokenizers/tokenizer.py]: use `TYPE_CHECKING` here?

File: src/megatron/bridge/data/hf_processors/squad.py (hunk 2 above)
Comment (cuichenx) [reply src/megatron/bridge/data/hf_processors/squad.py:63]: I think this is fine right? output can have special tokens. We'll need to modify the evaluation script to take these tokens into account. What do you suggest as the alternative?

File: src/megatron/bridge/data/hf_processors/squad.py (hunk 1 above)
Comment (cuichenx) [reply src/megatron/bridge/data/hf_processors/squad.py:60]: good point

File: src/megatron/bridge/data/hf_processors/squad.py (hunk 2 above)
Comment (cuichenx) [reply src/megatron/bridge/data/hf_processors/squad.py:63]: Ah I understand you now, yes we should process output as well. updated code
